Fix Airzone min/max climate temperatures

Noltari opened the pull request:
## Proposed change
Until now, the max/min of the current working mode (HEAT/COOL) would be set when starting Home Assistant, limiting the temperature range if the device mode was changed after that.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Noltari committed:
airzone: climate: fix max/min temps

The library now provides AZD_ABS_TEMP_MAX/AZD_ABS_TEMP_MIN which are useful for
devices with different max/min temperatures depending on the current working
mode (HEAT vs COOL).
These new values will have the highest/lowest max/min of both modes.
Until now, the max/min of the current working mode (HEAT/COOL) would be set
when starting Home Assistant, limiting the temperature range if the device
mode was changed after that.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into airzone-climate-max-min-temps

gjohansson-ST commented:
Should we not then update min/max every time mode is changed?
This seems not logical to have min/max which is not possible to use as the mode would perhaps not support it?

Noltari commented:
> Should we not then update min/max every time mode is changed? This seems not logical to have min/max which is not possible to use as the mode would perhaps not support it?

@gjohansson-ST is that supported by Homekit, Alexa, etc.?
I think that someone told me at some point that min/max values shouldn't be updated during the [Airzone climate platform review](https://github.com/home-assistant/core/pull/67924).
Not sure if it was @bdraco or someone else...

Noltari committed:
airzone: climate: update max/min temps

Some devices have different max/min climate temps depending on the active
mode (HEAT vs COOL), so we should update these values.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
> Should we not then update min/max every time mode is changed? This seems not logical to have min/max which is not possible to use as the mode would perhaps not support it?

@gjohansson-ST done in 988194d48614e2326a45b31ffcb8730b8172ac28

bdraco commented:
FWIW, homekit only checks min/max at the started event once
https://github.com/home-assistant/core/blob/0ce1117287f310cedeeef486366f704b0817ac4f/homeassistant/components/homekit/type_thermostats.py#L250

Noltari commented:
> FWIW, homekit only checks min/max at the started event once
> 
> https://github.com/home-assistant/core/blob/0ce1117287f310cedeeef486366f704b0817ac4f/homeassistant/components/homekit/type_thermostats.py#L250

Ok, so I will revert last commit then, thanks @bdraco!

Noltari committed:
Revert "airzone: climate: update max/min temps"

This reverts commit 988194d48614e2326a45b31ffcb8730b8172ac28.

gjohansson-ST commented:
In that case I think we need to check during service calls or it can possibly send a temp which is out of range of capability?

Noltari commented:
> In that case I think we need to check during service calls or it can possibly send a temp which is out of range of capability? Still seems to me a bit weird to handle it like this due to limiting factors in homekit.

Yes, I think that we need to check it and prevent sending an out of range temp, since from the tests I did, the API will seem to accept the new value...
```
$ curl -s --location --request PUT "http://192.168.1.25:3000/api/v1/hvac" -d '{"systemID": 1, "zoneID": 1, "setpoint": 35}' | jq
{
  "data": [
    {
      "systemID": 1,
      "zoneID": 1,
      "setpoint": 35
    }
  ]
}
```
But it didn't really accept it since it's out of range:
```
$ curl -s --location --request POST "http://192.168.1.25:3000/api/v1/hvac" -d '{"systemID": 1, "zoneID": 1}' | jq
{
  "data": [
    {
      "systemID": 1,
      "zoneID": 1,
      "name": "Salon",
      "thermos_type": 2,
      "thermos_firmware": "3.51",
      "thermos_radio": 0,
      "on": 0,
      "double_sp": 0,
      "coolsetpoint": 25,
      "coolmaxtemp": 30,
      "coolmintemp": 15,
      "heatsetpoint": 25,
      "heatmaxtemp": 30,
      "heatmintemp": 15,
      "maxTemp": 30,
      "minTemp": 15,
      "setpoint": 25,
      "roomTemp": 19,
      "sleep": 0,
      "temp_step": 0.5,
      "modes": [
        1,
        4,
        2,
        3,
        5
      ],
      "mode": 1,
      "speed": 0,
      "coldStage": 1,
      "heatStage": 1,
      "coldStages": 1,
      "heatStages": 1,
      "humidity": 32,
      "units": 0,
      "errors": [],
      "air_demand": 0,
      "floor_demand": 0,
      "cold_demand": 0,
      "heat_demand": 0,
      "heatangle": 0,
      "coldangle": 0,
      "master_zoneID": 1,
      "eco_adapt": "manual",
      "antifreeze": 0
    }
  ]
}
```

emontnemery commented:
> FWIW, homekit only checks min/max at the started event once
> 
> https://github.com/home-assistant/core/blob/0ce1117287f310cedeeef486366f704b0817ac4f/homeassistant/components/homekit/type_thermostats.py#L250

The home kit implementation can't be used as a source of truth for how entities should behave.

If we don't want to allow changing min/max, it should be documented here: https://developers.home-assistant.io/docs/core/entity/climate
If we do allow changing min/max, home kit needs to handle that the min/max values change.

Noltari commented:
> > FWIW, homekit only checks min/max at the started event once
> > https://github.com/home-assistant/core/blob/0ce1117287f310cedeeef486366f704b0817ac4f/homeassistant/components/homekit/type_thermostats.py#L250
> 
> The home kit implementation can't be used as a source of truth for how entities should behave.
> 
> If we don't want to allow changing min/max, it should be documented here: https://developers.home-assistant.io/docs/core/entity/climate If we do allow changing min/max, home kit needs to handle that the min/max values change.

Ok, in that case this PR isn't needed at all, so I'm gonna close it.

emontnemery commented:
@Noltari Why is this PR no longer needed, doesn't the PR set min/max to correct values?

Noltari commented:
> @Noltari Why is this PR no longer needed, doesn't the PR set min/max to correct values?

Because the Airzone API will update the max/min values when the mode is changed, and therefore AZD_TEMP_MAX and AZD_TEMP_MIN will change along with the mode.

gjohansson-ST commented:
> > @Noltari Why is this PR no longer needed, doesn't the PR set min/max to correct values?
> 
> Because the Airzone API will update the max/min values when the mode is changed, and therefore AZD_TEMP_MAX and AZD_TEMP_MIN will change along with the mode.

Currently `_attr_max_temp` and `_attr_min_temp` is only set when entity is created. Also think this PR should be reopened and these should be updated when the `hvac_mode` changes so we have correct min/max set.

Noltari committed:
Revert "Revert "airzone: climate: update max/min temps""

This reverts commit e4ead24f71cf2a16faec9253046419a59136daa1.

Noltari commented:
@gjohansson-ST @emontnemery `_attr_max_temp` and `_attr_min_temp` should now be updated with the coordinator.

gjohansson-ST commented on homeassistant/components/airzone/climate.py:
```diff
@@ -124,8 +124,6 @@ def __init__(
         self._attr_unique_id = f"{self._attr_unique_id}_{system_zone_id}"
         self._attr_supported_features = ClimateEntityFeature.TARGET_TEMPERATURE
         self._attr_target_temperature_step = API_TEMPERATURE_STEP
-        self._attr_max_temp = self.get_airzone_value(AZD_TEMP_MAX)
```
Think these needs to be set during init as well

Noltari replied:
> Think these needs to be set during init as well

_async_update_attrs is called at the end of the init function, so it's called:
https://github.com/home-assistant/core/blob/11857241544f8b445fd897a2ff9611fc9705fbb7/homeassistant/components/airzone/climate.py#L113-L139

gjohansson-ST replied:
Perhaps you can also add a test to see that it works as expected

Noltari replied:
> Perhaps you can also add a test to see that it works as expected

Sure, I can do that this afternoon :)

Noltari replied:
> Perhaps you can also add a test to see that it works as expected

Done in 9df40522fbc1e96691278d365cd15ae89dace2c6

Noltari commented:
Failure in coverage upload isn't related to changes:
https://github.com/home-assistant/core/actions/runs/5055638248/jobs/9072414053?pr=93161

Noltari committed:
tests: airzone: add max/min climate changes test

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into airzone-climate-max-min-temps

Noltari committed:
tests: airzone: fix dict copy

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

gjohansson-ST reviewed (approved):
LGTM 👍
